packages/@blockprotocol/graph/src/types/subgraph/edges/outward-edge.ts

```diff
@@ -20,18 +23,17 @@ type GenericOutwardEdge<K, E> = {
   rightEndpoint: E;
 };
 
+export type EntityIdAndTimestamp = {
+  baseId: EntityId;
+  timestamp: Timestamp;
+};
+
 export type OntologyOutwardEdge =
   | GenericOutwardEdge<OntologyEdgeKind, OntologyTypeEditionId>
   | GenericOutwardEdge<SharedEdgeKind, EntityEditionId>;
```

Comment:
Not introduced by this PR, but type implies you can have:
```ts
{
  kind: "IS_OF_TYPE",
  reversed: false,
  rightEndpoint: <some entity's id>
}
```
– is that right / known / fine?

Reply:
I did try looking at one that constrains the type properly but without explicit ways to define subtypes it's not really ergonomic :( 

(I want `satisfies` for types)

Reply:
Simple tweak that prevents someone constructing the incorrect example above
```ts
type GenericOutwardEdge<K, E, Reversed = boolean> = {
  kind: K;
  reversed: Reversed;
  rightEndpoint: E;
};

...

export type OntologyOutwardEdge =
  | GenericOutwardEdge<OntologyEdgeKind, OntologyTypeEditionId>
  | GenericOutwardEdge<SharedEdgeKind, EntityEditionId, true>;

export type KnowledgeGraphOutwardEdge =
  | GenericOutwardEdge<KnowledgeGraphEdgeKind, EntityIdAndTimestamp>
  | GenericOutwardEdge<SharedEdgeKind, OntologyTypeEditionId, false>;
```

Reply:
Obviously if `SharedEdgeKind` is expanded to include things that can go ontology<>knowledge then it'd have to be amended further. 

Probably at that point it would be worth introducing an separate record of which edges can go from ontology<>ontology, knowledge<>ontology (I think you just have to know this at the moment?)

Reply:
Bit concerned about adding loads of generics and decreasing the readability of the types, but I think generally this is a great suggestion, so I've added it. Thanks! (also added a few more generics 😅)

[Make GenericOutwardEdge even more generic](https://github.com/blockprotocol/blockprotocol/pull/820/commits/e5ed3aa96c043daec71842c3a0c6d1717d701456)